substra/sdk/backends/local/backend.py

```diff
@@ -0,0 +1,493 @@
+        testtuple_keys=None,
+    ):
+        tuple_count = functools.reduce(
+            lambda x, y: x + (len(y) if y else 0),
+            [
+                traintuple_keys,
+                composite_traintuple_keys,
+                aggregatetuple_keys,
+                testtuple_keys,
+            ],
+            0,
+        )
+        compute_plan = models.ComputePlan(
+            compute_plan_id=uuid.uuid4().hex,
+            status=models.Status.waiting.value,
+            traintuple_keys=traintuple_keys,
+            composite_traintuple_keys=composite_traintuple_keys,
+            aggregatetuple_keys=aggregatetuple_keys,
+            testtuple_keys=testtuple_keys,
+            id_to_key=dict(),
+            tag="",
+            tuple_count=tuple_count,
+            done_count=0,
+        )
+        return self._db.add(compute_plan)
+
+    def _add_traintuple(self, spec, exist_ok, spec_options=None):
+        # validation
+        algo = self._db.get(schemas.Type.Algo, spec.algo_key)
+        data_manager = self._db.get(schemas.Type.Dataset, spec.data_manager_key)
+        in_traintuples = (
+            [self._db.get(schemas.Type.Traintuple, key) for key in spec.in_models_keys]
+            if spec.in_models_keys is not None
+            else []
+        )
+
+        key_components = (
+            [_BACKEND_ID, spec.algo_key, spec.data_manager_key]
+            + spec.train_data_sample_keys
+            + spec.in_models_keys
+            if spec.in_models_keys is not None
+            else []
+        )
+        key = hasher.Hasher(values=key_components).compute()
+
+        # permissions
+        with_permissions = [algo, data_manager] + in_traintuples
+
+        authorized_ids = functools.reduce(
+            self.__intersect_permissions, with_permissions, None
+        )
+        public = authorized_ids is None
+        authorized_ids = list() if authorized_ids is None else list(authorized_ids)
+
+        # compute plan and rank
+        if not spec.compute_plan_id and spec.rank == 0:
+            #  Create a compute plan
+            compute_plan = self.__add_compute_plan(traintuple_keys=[key])
+            rank = 0
+            compute_plan_id = compute_plan.compute_plan_id
+        elif not spec.compute_plan_id and spec.rank is not None:
+            raise ValueError("Rank must be 0 to create compute plan.")
+        elif spec.compute_plan_id:
+            compute_plan = self._db.get(schemas.Type.ComputePlan, spec.compute_plan_id)
+            compute_plan_id = compute_plan.compute_plan_id
+            rank = (
+                len(compute_plan.traintuple_keys)
+                if compute_plan.traintuple_keys is not None
+                else 0 + len(compute_plan.composite_traintuple_keys)
+                if compute_plan.composite_traintuple_keys is not None
+                else 0 + len(compute_plan.aggregatetuple_keys)
+                if compute_plan.aggregatetuple_keys is not None
+                else 0
+            )
```

Comment:
I don't understand this code, could you explain it in comments?

My main issue is that for me, the rank of a traintuple is max of the in_models tuple ranks + 1 and I don't see that logic here. But maybe this way of calculating rank serves another purpose.

Reply:
My logic was that when we add a traintuple, there may already be aggregate tuples, composite tuples and traintuples, so the rank is the number of pre-existing tuples + 1
If we say that the rank is number of in_models + 1, what if one of the in_models is already in the compute plan with a rank > 0 ?

Can there be several tuples with the same rank in the compute plan ?

Reply:
my bad, ok for max of the in_models tuple ranks + 1

Reply:
my bad, updated !